src/common/KokkosKernels_SparseUtils.hpp

```diff
@@ -1955,6 +1963,167 @@ void kk_reduce_numrows_larger_than_threshold(
+                 const lno_t row_end_)
+      : Base(block_size_, row_begin_, row_end_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block(const lno_t col_idx) {
+    return Base::row_offset() + col_idx * Base::block_size;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block_stride() { return Base::size() * Base::block_size; }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type value(const lno_t col_idx, const lno_t block_row,
+                  const lno_t block_col) {
+    return block(col_idx) + block_row * block_stride() + block_col;
+  }
+};
+
+template <typename lno_t, typename size_type>
+class MatrixRowIndex<BSR, lno_t, size_type>
+    : public RowIndexBase<lno_t, size_type> {
+ public:
+  using Base = RowIndexBase<lno_t, size_type>;
+
+  KOKKOS_INLINE_FUNCTION
+  MatrixRowIndex(const lno_t block_size_, const lno_t row_begin_,
+                 const lno_t row_end_)
+      : Base(block_size_, row_begin_, row_end_) {}
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block(const lno_t col_idx) {
+    return Base::row_offset() + col_idx * Base::block_mtx_size();
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type block_stride() { return Base::block_size; }
+
+  KOKKOS_INLINE_FUNCTION
+  size_type value(const lno_t col_idx, const lno_t block_row,
+                  const lno_t block_col) {
+    return block(col_idx) + block_row * block_stride() + block_col;
+  }
+};
+
+template <typename mtx_t>
+struct MatrixTraits;
+
+template <typename scalar_t, typename lno_t, typename device,
+          typename mem_traits, typename size_type>
+struct MatrixTraits<
+    KokkosSparse::CrsMatrix<scalar_t, lno_t, device, mem_traits, size_type>> {
+  static constexpr auto format = KokkosKernels::BlockCRS;
+};
+
+template <typename scalar_t, typename lno_t, typename device,
+          typename mem_traits, typename size_type>
+struct MatrixTraits<KokkosSparse::Experimental::BsrMatrix<
+    scalar_t, lno_t, device, mem_traits, size_type>> {
+  static constexpr auto format = KokkosKernels::BSR;
+};
+
+template <SparseMatrixFormat /* outFormat */>
+struct MatrixConverter;
+
+template <>
+struct MatrixConverter<BlockCRS> {
+  template <
+      typename scalar_t, typename lno_t, typename device, typename size_type,
+      typename crsMat_t =
+          KokkosSparse::CrsMatrix<scalar_t, lno_t, device, void, size_type>>
+  static crsMat_t from_crs(
+      const std::string &label,
+      const KokkosSparse::CrsMatrix<scalar_t, lno_t, device, void, size_type>
+          &mtx,
+      lno_t block_size) {
+    using graph_t        = typename crsMat_t::StaticCrsGraphType;
+    using lno_view_t     = typename graph_t::row_map_type::non_const_type;
+    using lno_nnz_view_t = typename graph_t::entries_type::non_const_type;
+    using scalar_view_t  = typename crsMat_t::values_type::non_const_type;
+    lno_view_t rows;
+    lno_nnz_view_t entries;
+    scalar_view_t vals;
+    size_t num_rows, num_cols;
+
+    KokkosKernels::Impl::kk_create_blockcrs_from_blockcrs_formatted_point_crs(
```

Comment:
Does ``kk_create_blockcrs_from_blockcrs_formatted_point_crs`` assume that the matrix already has block structure? If so, is there a way to convert a general CrsMatrix to BlockCrs?

Reply:
No, `kk_create_blockcrs_from_blockcrs_formatted_point_crs()` here in the test it takes output from `kk_generate_diagonally_dominant_sparse_matrix()`, which doesn't take block size as argument, so I don't see block format assumed/required.

`BlockCrsMatrix` has a constructor that takes `CrsMatrix` and block size to be imposed - I guess we could just replace  `kk_create_blockcrs_from_blockcrs_formatted_point_crs()` with it.

Reply:
OK. And yes, it would be good for ``kk_create_blockcrs_from_blockcrs_formatted_point_crs`` to not duplicate that BlockCrs constructor (assuming they have the same behavior).

Reply:
There are two methods, both take and return CRS sparse views:
1. `kk_create_blockcrs_formated_point_crsmatrix()` takes plain CRS matrix and creates dense blocks filling in zeros - but _without_ changing matrix dimensions - outputting "semi-block CRS" matrix.
2. `kk_create_blockcrs_from_blockcrs_formatted_point_crs()` takes "semi-block CRS" with dense blocks and reduces matrix dimensions to `(num_rows / block_size, num_cols / block_size)`, so it does assume/require (apologies for posting misleading information in the previous post 🤦‍♂️)

I believe `BlockCrsMatrix( crs_mtx, block_size )` should be able replace either second method alone or the combination of both.  
@uhetmaniuk Would you be able to confirm if that is actually a good replacement ? I feel like I can very easily miss on implementation details relevant for the above statement...  

> **Note:** the only other place where this workflow is used is [block PCG perf test](https://github.com/kokkos/kokkos-kernels/blob/master/perf_test/sparse/KokkosSparse_block_pcg.cpp#L346-L383) which, unlike GS, also runs the kernel on the intermediate semi-block CRS matrix.

> **Note:** I cleaned unnecessary intermediate `kk_create_blockcrs_formated_point_crsmatrix()` calls in block GS tests to on 1de48fcd86748fb618317c841d5418e94576ff1b

Reply:
I found following issues in the deeper analysis:
* `BlockCrsMatrix( crs_mat, block_size )` constructor **requires block formatted** CRS input - it does not adjust values.
* GS unit test uses `kk_create_blockcrs_formated_point_crsmatrix()` to convert input of 500 rows to 72 blocks, 7x7 each, expanding the matrix to 504 scalar rows, filling with zeros if needed, but:
  - `BlockCrsMatrix( crs_mat, block_size )` cuts blocks down, producing 71 block rows / 497 scalar rows;
  - `BsrMatrix( crs_mat, block_size )` does not support block size which does not divide matrix size (runtime error);
* `create_random_y_vector()` and `create_random_y_vector_mv()` do not work with block matrices yet - just tried to add that quickly and failed on linker errors in multi-vector `spmv()` for `BlockCrsMatrix`...

Would it be possible for us to discuss and sort out overall block CRS matrix generation - with related utilities - on separate issue/PR ?